File: public/codes/2-Regresion/1-Simple_Linear_Regression/simple_linear_regression.py
# -*- coding: utf-8 -*-

# --------------------Regresion lineal simple--------------------

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# =============================================================================
# --------------------Importando dataset--------------------
# =============================================================================

# Dataset de empleados que incluye los años de esperiencia de un empleado
# y de cuanto es su salario.
# Columnas: |YearExperience| (var independiente)
#           |Salary (var. dependiente)
dataset = pd.read_csv('Salary_Data.csv')
# Se toman las primeras columnas, menos la ultima.
X = dataset.iloc[:,:-1].values 
# Se toma la ultima columna.
y = dataset.iloc[:,1].values

# =============================================================================
# --------------------Dividir el dataset en conjunto--------------------
# --------------------de entrenamiento y conjunto de testing--------------------
# =============================================================================

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=1/3,random_state=2)

# =============================================================================
# --------------------Crear modelo de Regresion Lineal--------------------
# --------------------Simple con el conjunto de entrenamiento--------------------
# =============================================================================
from sklearn.linear_model import LinearRegression

# Modulo que crea los modelos de regresion lineal
regression = LinearRegression()
# (Conjunto de variables independientes, Conjunto de variables a predecir)
regression.fit(X_train ,y_train)
# =============================================================================
# --------------------Predecir el conjunto de test--------------------
# =============================================================================
# y_pred nos entrega el modelo lineal, en base al conjunto de testing.
# Al visualizar ambas tablas los valores deberian de ser casi similares.
y_pred = regression.predict(X_test)

# =============================================================================
# --------------------Visualizar los resultados de entrenamiento--------------------
# =============================================================================
# Para mostrar la grafica de datos de entrenamiento y la recta del modelo lineal

# Dubujando los puntos del conjunto de entrenamiento de color rojo.
plt.scatter(X_train,y_train, color="red")
# Dobujando la linea de regresion de color azul
plt.plot(X_train,regression.predict(X_train),color="blue")
# Dando etiquetas a la grafica
plt.title("Sueldo vs Años de experiencia(Entrenamiento)")
plt.ylabel("Sueldo ($)")
plt.xlabel("Anos de experiencia")
plt.show()

# =============================================================================
# --------------------Visualizar los resultados de test--------------------
# =============================================================================
# Para mostrar la grafica de datos de testing y la recta del modelo lineal

# Dubujando los puntos del conjunto de testing de color rojo.
plt.scatter(X_test,y_test, color="red")
# Dobujando la linea de regresion de color azul
plt.plot(X_train,regression.predict(X_train),color="blue")
# Dando etiquetas a la grafica
plt.title("Sueldo vs Años de experiencia(Testing)")
plt.ylabel("Sueldo ($)")
plt.xlabel("Anos de experiencia")
plt.show()

File: public/codes/2-Regresion/1-Simple_Linear_Regression/simple_linear_regression.R
# Regresion Lineal Simple

# =============================================================================
# --------------------Importando dataset--------------------
# =============================================================================
# Dataset de empleados que incluye los años de esperiencia de un empleado
# y de cuanto es su salario.
# Columnas: |YearExperience| (var independiente)
#           |Salary (var. dependiente)
dataset = read.csv('Salary_Data.csv')

# =============================================================================
# --------------------Dividir el dataset en conjunto--------------------
# --------------------de entrenamiento y conjunto de testing--------------------
# =============================================================================

# install.packages("caTools") # solo se necesita ejecutar una vez
# library(caTools)

# configurando semilla aleatoria para la division de datos
set.seed(5)
# se elige el porcentaje de los datos para el training en %
# se selecciona la columna de la variable dependiente o a predecir
split = sample.split(dataset$Salary,SplitRatio = 2/3)
print(split)
# Dividiendo el conjunto , True para el training
training_set = subset(dataset,split == TRUE)
# Dividiendo el conjunto , False para el test
testing_set = subset(dataset,split == FALSE)

# =============================================================================
# --------------------Crear modelo de Regresion Lineal--------------------
# --------------------Simple con el conjunto de entrenamiento--------------------
# =============================================================================

regressor=lm(formula=Salary ~ YearsExperience, data=training_set)
summary(regressor)

# =============================================================================
# --------------------Predecir el conjunto de test--------------------
# =============================================================================

y_pred=predict(regressor,newdata=testing_set)
print(y_pred)

# =============================================================================
# ---------------Visualizar los resultados de entrenamiento--------------------
# =============================================================================
# Para mostrar la grafica de datos de entrenamiento y la recta del modelo lineal

# Agregando componentes a mostrar
ggplot()+
  # Dobujando los puntos de entrenamiento
  geom_point(aes(x=training_set$YearsExperience,y=training_set$Salary),
             colour="red")+
  # Dobujando la linea de la prediccion, en base al entrenamiento
  geom_line(aes(x=training_set$YearsExperience,y=predict(regressor,newdata=training_set)),
            colour="blue")+
  ggtitle("Sueldo vs Años de experiencia(Entrenamiento)")+
  xlab("Sueldo ($)")+
  ylab("Anos de experiencia")

# =============================================================================
# --------------------Visualizar los resultados de test--------------------
# =============================================================================
# Para mostrar la grafica de datos de testing y la recta del modelo lineal


# Agregando componentes a mostrar
ggplot()+
  # Dobujando los puntos de testing
  geom_point(aes(x=testing_set$YearsExperience,y=testing_set$Salary),
             colour="red")+
  # Dobujando la linea de la prediccion, en base al testing
  geom_line(aes(x=testing_set$YearsExperience,y=y_pred),
            colour="blue")+
  ggtitle("Sueldo vs Años de experiencia(testing)")+
  xlab("Sueldo ($)")+
  ylab("Anos de experiencia")



